Restructure to adhere closer to Ruby Gem conventions and avoid clashes

twometresteve opened the pull request:
## Goal

Restructure the repo to adhere more closely to Ruby Gem conventions and avoid collisions in the default namespace.

## Design

The diffs here look quite scary, with most files touched in some way.  It will probably be easier to review by checking out the branch locally to verify that the overall end state makes sense, rather than trying to consider each delta individually.

## Changeset

- `bugsnag-maze-runner` entry point has been removed so only `bundle exec maze-runner`can be used.  Having both was deemed unnecessary and this avoids annoying duplicate code warning in an intelligent IDE.
- `MazeRunner` class containing the `config`, `hooks`, etc is renamed renamed to just `Maze`.
- Core files have moved from `lib/features/support` into `lib/maze` and classes placed in the `Maze` namespace.
- Some Rubocop violations resolved along the way.

## Tests

The unit and e2e tests caught a great deal of issues where namespace qualification was missing or incorrect, giving me a good level of confidence in the overall change.  But we can also afford to take on more risk in the change, as notifiers will be upgraded to use the new version in turn and we can iron our any wrinkles as we find them.

imjoehaines commented on lib/maze/option.rb:
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+# Provides the set of Maze Runner command line options
+module Maze
+  class Option
```
Would making this a `Maze::Option` module, then renaming the classes to `Maze::Option::Parser`, `Maze::Option::Processor` & `Maze::Option::Validator` make sense?

twometresteve replied:
Yes, completely - done.

imjoehaines commented on lib/maze/options/option_validator.rb:
```diff
@@ -0,0 +1,103 @@
+        end
+
+        begin
+          JSON.parse(options[Option::CAPABILITIES])
+        rescue JSON::ParserError
+          errors << "--#{Maze::Option::CAPABILITIES} must be valid JSON (given #{options[Option::CAPABILITIES]})"
```
Some places in this file are using `Option::<name>` but others do `Maze::Option::<name>`

twometresteve replied:
All changed to the shorter form

imjoehaines commented on lib/maze/drivers/resilient_driver.rb:
```diff
@@ -0,0 +1,51 @@
+      end
+
+      def respond_to_missing?(method_name, include_private = false)
+        @driver.respond_to_missing? method_name
```
```suggestion
        @driver.respond_to_missing? method_name, include_private
```

twometresteve replied:
Fixed

twometresteve commented on lib/maze.rb:
```diff
@@ -1,19 +1,22 @@
 # providing an alternative to the proliferation of global variables or singletons.
-class MazeRunner
+module Maze
+  VERSION ||= '4.0.0'
```
You get a load of warnings come out at runtime about constants already being declared without it.  It seemed like a common approach to take from having Googled it, and we'd already done the same in `Devices`.

twometresteve replied:
Now resolved thanks to your later comment.

twometresteve commented on lib/maze/capabilities/devices.rb:
```diff
@@ -0,0 +1,113 @@
+    # Provides a source of capabilities used to run tests against specific BrowserStack devices
+    # noinspection RubyStringKeysInHashInspection
+    class Devices
+      APPIUM_1_6_5 ||= '1.6.5'
+      APPIUM_1_7_2 ||= '1.7.2'
+      APPIUM_1_8_0 ||= '1.8.0'
+      APPIUM_1_15_0 ||= '1.15.0'
```
Ah, didn't see this when I wrote my first response.  These were put in some time ago and the diff here is just misleading.  I'm not sure if there is another or better way of doing it and it's what I've seen on various SO posts - open to other suggestions, of course.

imjoehaines replied:
I think the problem was Cucumber requiring these files — https://github.com/bugsnag/maze-runner/blob/fd8ce09c7ab913afab69bf3fe17a48a1613254e3/bin/maze-runner#L64-L66

We should be able to remove `lib/maze` from the list of paths and load those paths up-front if necessary (either in `bin/maze-runner` or `lib/maze.rb`)

twometresteve replied:
Ah... great!  This is starting to come together now.  I've removed those lines, added a load of `require`s at the top of `bin/maze-runner` and moved `hooks.rb` back under features (because it's Cucumber specific).

There's probably still room for improvement, but it's no longer necessary to use `||=`.

twometresteve commented on lib/maze/docker.rb:
```diff
@@ -0,0 +1,108 @@
+        @services_started = false
+      end
+
+      def compose_project_name
+        @compose_project_name ||= nil
+      end
+
+      attr_writer :compose_project_name
```
I don't fully understand what's going on, but I _think_ it's needed because the class uses `class << self`.  If I remove the setter and change `compose_project_name` to `@compose_project_name` on L92 it falls over in a heap.

twometresteve commented on lib/maze/logger.rb:
```diff
@@ -0,0 +1,71 @@
+# Logger classes
+module Maze
+  # A logger, with level configured according to the environment
+  class MazeLogger < Logger
```
Done.

twometresteve commented on lib/maze/hooks/client_hooks.rb:
```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module Maze
+  module Hooks
+    # Provides the ability for callbacks to be provided as part of running Cucumber.
+    # These are akin to Cucumber's AfterConfiguration, Before and After hooks, but are invoked in such a way that
+    # Maze Runner's hooks do not interfere with callbacks registered by clients.
+    class ClientHooks
```
Just as I started saying to myself "I'm surprised Rubocop didn't suggest this to me", it actually wants the nested rather than compact style.  I guess it's part of wider conversation that we can have about style one day, but I'll leave it as is for now (although I have now renamed the file to match the class name).

twometresteve commented on lib/maze/network.rb:
```diff
@@ -0,0 +1,49 @@
+
+module Maze
+  # Sets the maximum number of times Maze runner will ping to see if a port is open, defaulting to 100
+  MAX_MAZE_CONNECT_ATTEMPTS = ENV.include?('MAX_MAZE_CONNECT_ATTEMPTS')? ENV['MAX_MAZE_CONNECT_ATTEMPTS'].to_i : 100
```
Changed

twometresteve commented on lib/maze/assertions/request_set_assertions.rb:
```diff
@@ -0,0 +1,95 @@
+require 'test/unit'
+require_relative '../helper'
+
+module Maze
+  module Assertions
```
See previous response, Rubocop no-likey.

twometresteve commented on lib/maze/option.rb:
```diff
@@ -0,0 +1,29 @@
+module Maze
+  class Option
+    # Common options
+    SEPARATE_SESSIONS ||= 'separate-sessions'
```
Not any more. :-)

twometresteve commented on test/appium_driver_test.rb:
```diff
@@ -1,9 +1,9 @@
+require_relative '../lib/maze/drivers/appium_driver'
 
-class AppiumDriverTest < Test::Unit::TestCase
+class Maze::Drivers::AppiumDriverTest < Test::Unit::TestCase
```
Fixed - caused by a global find/replace.

twometresteve commented on lib/maze/hooks/hooks.rb:
```diff
@@ -9,7 +9,7 @@
 
```
You're quite right - have moved it back.

twometresteve commented on lib/maze/compare.rb:
```diff
@@ -0,0 +1,160 @@
+
+module Maze
+  # Provides a way to delivering results of element comparisons to test steps
+  class CompareResult
```
Have now changed to a module of `Compare` and renamed the methods with it, so now it's `Maze::Compare.value`, etc.

twometresteve commented on lib/maze/capabilities/capabilities.rb:
```diff
@@ -0,0 +1,74 @@
+module Maze
+  module Capabilities
+    # Appium capabilities for each target farm
+    class Capabilities
```
No, I don't think it adds much at all and the grouping of devices and capabilities together was clumsy.  I've now removed the `Maze::Capabilities` module entirely, with its contents simply moving up into `Maze`.

twometresteve commented on lib/maze/drivers/appium_driver.rb:
```diff
@@ -0,0 +1,142 @@
+module Maze
+  module Drivers
+    # Provide a thin layer of abstraction above @see Appium::Driver
+    class AppiumDriver < Appium::Driver
```
Agreed - changed.

twometresteve commented on lib/maze/helper.rb:
```diff
@@ -0,0 +1,62 @@
+
+# A collection of helper routines
+module Maze
+  class << self
```
I think prefer creating a module for them, as there will come a point that we'll want to refactor what is really a holding area for things that don't have a logical home yet.  Putting helper routines in their own module should make them easier to identify. 
I've gone for the singular `Helper` here, with the name referring more to the module than the functions in it, but I'd be happy to change to plural if it was more conventional.

twometresteve commented:
As part of actioning the review comments I realised that any unit tests in a folder were not being run on CI due to the path in the `Rakefile`.  I've corrected that now and made a further series of unit test and code updates so that all the unit tests pass.  

Some of the integration tests have been overtaken by events (using step names right back from v1 and simulators, which we no longer use).  I've deleted those that no longer apply to reduce the noise level and make it easier to see where there are test gaps, which we're looking to cover over time anyway.  The interactive cli tests do do seem to run on CI for some reason, so have not added them to CI at this stage.
